src/Microsoft.OData.Core/UriParser/Binders/MetadataBindingUtils.cs

```diff
@@ -58,19 +58,31 @@ internal static SingleValueNode ConvertToTypeIfNeeded(SingleValueNode source, IE
                 }
 
                 ConstantNode constantNode = source as ConstantNode;
-                if (constantNode != null && constantNode.Value != null && source.TypeReference.IsString() && targetTypeReference.IsEnum())
+                // Check if the source node is a constant node, not null, and the source type is either string or integral
+                // and the target type is an enum.
+                if (constantNode != null && constantNode.Value != null && (source.TypeReference.IsString() || source.TypeReference.IsIntegral()) && targetTypeReference.IsEnum())
                 {
                     string memberName = constantNode.Value.ToString();
                     IEdmEnumType enumType = targetTypeReference.Definition as IEdmEnumType;
                     if (enumType.Members.Any(m => string.Compare(m.Name, memberName, StringComparison.Ordinal) == 0))
                     {
                         string literalText = ODataUriUtils.ConvertToUriLiteral(constantNode.Value, default(ODataVersion));
-                        return new ConstantNode(new ODataEnumValue(constantNode.Value.ToString(), targetTypeReference.Definition.ToString()), literalText, targetTypeReference);
+                        return new ConstantNode(new ODataEnumValue(memberName, enumType.ToString()), literalText, targetTypeReference);
                     }
-                    else
+
+                    // If the member name is an integral value, we should try to convert it to the enum member name.
+                    if (int.TryParse(memberName, out int memberIntegralValue))
                     {
-                        throw new ODataException(ODataErrorStrings.Binder_IsNotValidEnumConstant(memberName));
+                        // Find the enum member with the matching integral value
+                        var enumMember = enumType.Members.FirstOrDefault(m => m.Value.Value == memberIntegralValue);

```

Comment:
Across this library's source code, it's the practice to use actual type name instead of `var`

Reply:
@gathogojr 

Thanks for this John. I have replaced the `var `types with explicit types